### PR TITLE
Feat/adding unittests for k8s prereq module

### PR DESCRIPTION
**Pull request by boekhorstb1**
- added unittests: note that these only run with pytest currently. Try:  `.venv/bin/python -m pytest tests/modules/test_k8s_prerequisites_check.py`
- when running with unittest there are various import errors. Try: `/app/rookify/.venv/bin/python -m unittest /app/rookify/src/tests/modules/test_k8s_prerequisites_check.py`, also note that these import errors do not occure when testing `ExampleHandler` module, yet import is done in the same way. Try: `/app/rookify/.venv/bin/python -m unittest /app/rookify/src/tests/modules/test_example.py`
- because the ModuleHandler loads all the module, I found that I had to mock the class `TestK8sPrerequisitesCheckHandler` itself, thereby keping the code for preflight method the same... maybe you know another way to mock the ModuleHandler before it gets loaded?

### CODE REVIEW

**Review by NotTheEvilOne — APPROVED** [review]
`pytest` is required for unit testing. `pyproject.toml` contains configuration to make it work as expected which `unittest` to not support (even by command line argument). Later is at least true for 3.12.3 of Python at the time of writing.

`load_modules()` is not mockable at the moment as it scans through the `rookify.modules` directory to identify modules to be loaded.

**Comment by boekhorstb1** [comment]
forgot to merge this, now have to repair conflicts :)

**Comment by boekhorstb1** [comment]
ok repaired it and added another small unittest

**Comment by boekhorstb1** [comment]
Ok I noted an issue: there seems to be some discrepancy between the local pre-comit and the pre-commit run in github actions. I now ran  git commit --no-verify to avoid the local pre-commit throwing an error: github actions seems to run through everything fine though

File: Makefile [hunk 1]
```diff
@@ -28,17 +28,16 @@ help: ## Display this help message
 	    awk -F ':.*?## ' 'NF==2 {printf "  %-26s%s\n\n", $$1, "${COLOUR_GREEN}"$$2"${COLOUR_END}"}'
 
 .PHONY: setup
-setup: setup-pre-commit check-radoslib setup-venv ## Setup the pre-commit environment and then the venv environment
+setup: check-radoslib setup-venv setup-pre-commit ## Setup the pre-commit environment and then the venv environment
 
 .PHONY: setup-pre-commit
 setup-pre-commit:
-	pip install --user pre-commit && pre-commit install
+	./.venv/bin/pip install --user pre-commit && ./.venv/bin/python3.12 -m pre_commit install
```
Comment (NotTheEvilOne) [inline Makefile:36]: virtualenv access should not be version dependent

File: Makefile [hunk 2]
```diff
@@ -28,17 +28,16 @@ help: ## Display this help message
...
 
 .PHONY: setup-venv
 setup-venv:
-	python -m venv --system-site-packages ./.venv && \
-	source ./.venv/bin/activate && \
-	pip install -r requirements.txt
+	python3 -m venv --system-site-packages ./.venv && \
```
Comment (NotTheEvilOne) [inline Makefile:39]: We should not replace `python` with `python3` here as `python2` isn't a future proof version anyway.

File: Makefile [hunk 3]
```diff
@@ -68,7 +67,7 @@ check-radoslib: ## Checks if radoslib is installed and if it contains the right
 run-local-rookify: ## Runs rookify in the local development environment (requires setup-venv)
 	$(eval PYTHONPATH="${PYTHONPATH}:$(pwd)/src")
 	source ./.venv/bin/activate && \
-	cd src && python -m rookify
+	cd src && python3 -m rookify
```
Comment (NotTheEvilOne) [inline Makefile:70]: Same here for `python3` replacement

File: tests/modules/test_k8s_prerequisites_check.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        # No response option
+        self.empty_response = False
+
+    def _request_callback(
+        self, method: str, *args: List[Any], **kwargs: Dict[Any, Any]
+    ) -> Any:
+        if method == "apps_v1_api.list_deployment_for_all_namespaces":
+            if self.empty_response is True:
+                return MockResponse([])
+            return MockResponse(["apple", "banana", "cherry"])
```
Comment (NotTheEvilOne) [inline tests/modules/test_k8s_prerequisites_check.py:23]: I wonder if there's no equivalent available with a more intuitive name for `MockResponse`? See [NamespaceList](https://github.com/kubernetes-client/python/blob/041103952a93d1fa7a4478f972b581374f953ae6/kubernetes/docs/V1NamespaceList.md?plain=1). E.g. [DeploymentList](https://github.com/kubernetes-client/python/blob/041103952a93d1fa7a4478f972b581374f953ae6/kubernetes/client/models/v1_deployment_list.py)?

File: tests/modules/test_k8s_prerequisites_check.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        # Mock configuration
+        self.config = {"rook": {"cluster": {"namespace": "test-namespace"}}}
+        # No response option
+        self.empty_response = False
+
+    def _request_callback(
+        self, method: str, *args: List[Any], **kwargs: Dict[Any, Any]
+    ) -> Any:
+        if method == "apps_v1_api.list_deployment_for_all_namespaces":
+            if self.empty_response is True:
```
Comment (NotTheEvilOne) [inline tests/modules/test_k8s_prerequisites_check.py:21]: You may want to use a more descriptive name for this flag to differentiate test calls for the deployment list.

File: tests/modules/test_k8s_prerequisites_check.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                return MockResponse([])
+            return MockResponse(["apple", "banana", "cherry"])
+        if method == "core_v1_api.list_namespace":
+            return self._mock_list_namespace()
+
+    def _mock_list_namespace(self) -> Any:
```
Comment (NotTheEvilOne) [inline tests/modules/test_k8s_prerequisites_check.py:27]: Same here. You can reuse Python `kubernetes-client` objects for that purpose :)

**Review by NotTheEvilOne — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (boekhorstb1) [reply Makefile:36]: but the gitlab ci is version dependent and it is running on python3.12. So should we then make ci dependent on a python version? Should rookify itself not be dependent on a specific python version?

File: Makefile (hunk 2 above)
Comment (boekhorstb1) [reply Makefile:39]: ok, I thought ubuntu did not have the python command but only python3, yet in venv I now see there is a python slimlink to python3 :) so I can change that back

File: tests/modules/test_k8s_prerequisites_check.py (hunk 4 above)
Comment (boekhorstb1) [reply tests/modules/test_k8s_prerequisites_check.py]: ok changed name to 'V1DeploymentList'

File: tests/modules/test_k8s_prerequisites_check.py (hunk 6 above)
Comment (boekhorstb1) [reply tests/modules/test_k8s_prerequisites_check.py]: ok thanks! good idea, used kubernets-client namespace options to create the mock now

File: Makefile (hunk 1 above)
Comment (boekhorstb1) [reply Makefile:47]: ok I now added some dynamic python check to add the python executable (ubuntu uses `python3` or `python2` not `python` unless you install the package `python-is-python3` haha

File: tests/modules/test_k8s_prerequisites_check.py (hunk 5 above)
Comment (boekhorstb1) [reply tests/modules/test_k8s_prerequisites_check.py]: thanks good idea, changed flag to empty_deployment_list

**Review by NotTheEvilOne — APPROVED** [review]
LGTM

**Pull request merged**